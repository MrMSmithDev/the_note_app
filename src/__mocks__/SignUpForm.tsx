import { SignUpFormProps } from '@components/SignUpForm';

export default function SignUpForm({ closeForm, toggleForm }: SignUpFormProps) {
  return (
    <div data-testid="mock-sign-up-form">
      <button onClick={closeForm}>close sign up</button>
      <button onClick={toggleForm}>switch to sign in form</button>
    </div>
  );
}
