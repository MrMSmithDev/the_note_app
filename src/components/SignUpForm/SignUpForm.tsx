import { EmailIcon, HidePasswordIcon, ShowPasswordIcon, UserIcon } from '@components/icons';
import React, { useState } from 'react';

export interface SignUpFormProps {
  closeForm: () => void;
  toggleForm: () => void;
}

const SignUpForm: React.FC<SignUpFormProps> = ({ closeForm, toggleForm }) => {
  const [username, setUsername] = useState<string>('');
  const [email, setEmail] = useState<string>('');
  const [password, setPassword] = useState<string>('');
  const [repeatPassword, setRepeatPassword] = useState<string>('');

  const [revealPassword, setRevealPassword] = useState<boolean>(false);
  const [revealRepeatPassword, setRevealRepeatPassword] =
    useState<boolean>(false);

  function handleUsernameChange(e: React.ChangeEvent<HTMLInputElement>) {
    setUsername(e.currentTarget.value);
  }

  function handleEmailChange(e: React.ChangeEvent<HTMLInputElement>) {
    setEmail(e.currentTarget.value);
  }

  function handlePasswordChange(e: React.ChangeEvent<HTMLInputElement>) {
    setPassword(e.currentTarget.value);
  }

  function handleRepeatPasswordChange(e: React.ChangeEvent<HTMLInputElement>) {
    setRepeatPassword(e.currentTarget.value);
  }

  function toggleRevealPassword(e: React.MouseEvent<HTMLButtonElement>) {
    e.preventDefault();
    setRevealPassword((prev) => !prev);
  }

  function toggleRevealRepeatPassword(e: React.MouseEvent<HTMLButtonElement>) {
    e.preventDefault();
    setRevealRepeatPassword((prev) => !prev);
  }

  function switchForm(e: React.MouseEvent<HTMLButtonElement>) {
    e.preventDefault();
    toggleForm();
  }

  function submitForm(e: React.FormEvent<HTMLButtonElement>) {
    e.preventDefault();

    console.log(username);
    console.log(email);
    console.log(password);
    console.log(repeatPassword);

    closeForm();
  }

  return (
    <form className="flex flex-col gap-5 bg-background dark:bg-background-dark rounded-md p-5">
      <h3 className="font-kanit font-bold text-2xl text-transparent bg-clip-text mr-auto bg-linear-to-tr from-blue-400 to-blue-600 py-2 select-none w-full text-center">
        Sign up
      </h3>
      <div className="relative">
        <UserIcon className="absolute h-5 w-5 top-2 left-2" />
        <label className="sr-only" htmlFor="username"></label>
        <input
          className="w-full p-2 rounded border-gray-200 border-2 border-solid dark:border-gray-800 focus:outline-gray-300 text-gray-800 dark:text-gray-200 tracking-wide pl-8"
          type="text"
          name="username"
          id="username"
          placeholder="Username"
          value={username}
          onChange={handleUsernameChange}
        />
      </div>
      <div className="relative">
        <EmailIcon className="absolute h-5 w-5 top-2 left-2" />
        <label className="sr-only" htmlFor="email"></label>
        <input
          className="w-full p-2 rounded border-gray-200 border-2 border-solid dark:border-gray-800 focus:outline-gray-300 text-gray-800 dark:text-gray-200 tracking-wide pl-8"
          type="email"
          name="email"
          id="email"
          placeholder="Email@example.com"
          value={email}
          onChange={handleEmailChange}
        />
      </div>
      <div className="relative">
        <button
          onClick={toggleRevealPassword}
          className="absolute h-5 w-5 top-2 left-2 cursor-pointer"
        >
          <span className="sr-only">show password</span>
          {revealPassword ? <ShowPasswordIcon /> : <HidePasswordIcon />}
        </button>
        <label className="sr-only" htmlFor="password"></label>
        <input
          className="w-full p-2 rounded border-gray-200 border-2 border-solid dark:border-gray-800 focus:outline-gray-300 text-gray-800 dark:text-gray-200 tracking-wide pl-8"
          type={revealPassword ? 'text' : 'password'}
          name="password"
          id="password"
          placeholder="Password"
          value={password}
          onChange={handlePasswordChange}
        />
      </div>
      <div className="relative">
        <button
          onClick={toggleRevealRepeatPassword}
          className="absolute h-5 w-5 top-2 left-2 cursor-pointer"
        >
          <span className="sr-only">show repeat password</span>
          {revealRepeatPassword ? <ShowPasswordIcon /> : <HidePasswordIcon />}
        </button>
        <label className="sr-only" htmlFor="repeatPassword"></label>
        <input
          className="w-full p-2 rounded border-gray-200 border-2 border-solid dark:border-gray-800 focus:outline-gray-300 text-gray-800 dark:text-gray-200 tracking-wide pl-8"
          type={revealRepeatPassword ? 'text' : 'password'}
          name="repeatPassword"
          id="repeatPassword"
          placeholder="Repeat password"
          value={repeatPassword}
          onChange={handleRepeatPasswordChange}
        />
      </div>
      <button
        type="submit"
        className="py-2 px-4 rounded-md font-quicksand font-bold text-gray-100 mr-auto bg-linear-to-tr from-blue-400 to-blue-600 w-full cursor-pointer shadow"
        onClick={submitForm}
      >
        Sign up
      </button>
      <div className="text-center">
        <p className="text-gray-800 dark:text-gray-200 text-xs">
          Already have an account?
        </p>
        <button
          onClick={switchForm}
          className="text-blue-400 text-sm font-bold cursor-pointer"
        >
          Sign in
        </button>
      </div>
    </form>
  );
};

export default SignUpForm;
