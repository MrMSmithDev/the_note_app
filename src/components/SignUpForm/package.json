{
  "name": "sign-up-form",
  "version": "1.0.0",
  "main": "SignUpForm.tsx",
  "dependencies": {
    "react": "^19.0.0"
  }
}
