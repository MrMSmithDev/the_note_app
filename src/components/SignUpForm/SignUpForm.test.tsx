import { screen, render, fireEvent } from '@testing-library/react';
import SignUpForm from './SignUpForm';

describe('SignUpForm component', () => {
  const mockCloseForm = jest.fn();
  const mockToggleForm = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders to match snapshot', () => {
    const { asFragment } = render(
      <SignUpForm closeForm={mockCloseForm} toggleForm={mockToggleForm} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly', () => {
    render(
      <SignUpForm closeForm={mockCloseForm} toggleForm={mockToggleForm} />
    );

    expect(screen.getAllByText('Sign up').length).toBe(2);
    expect(screen.getByPlaceholderText('Username')).toBeInTheDocument();
    expect(
      screen.getByPlaceholderText('Email@example.com')
    ).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Password')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Repeat password')).toBeInTheDocument();
    expect(screen.getByTestId('user-icon')).toBeInTheDocument();
    expect(screen.getByTestId('email-icon')).toBeInTheDocument();
    expect(screen.getAllByTestId('hide-password-icon').length).toBe(2);
    expect(screen.getByRole('button', { name: 'Sign up' })).toBeInTheDocument();
  });

  it('updates username and email fields', () => {
    render(
      <SignUpForm closeForm={mockCloseForm} toggleForm={mockToggleForm} />
    );

    const usernameInput = screen.getByPlaceholderText(
      'Username'
    ) as HTMLInputElement;
    const emailInput = screen.getByPlaceholderText(
      'Email@example.com'
    ) as HTMLInputElement;

    fireEvent.change(usernameInput, { target: { value: 'test_username' } });
    fireEvent.change(emailInput, { target: { value: 'test@email.com ' } });

    expect(usernameInput.value).toBe('test_username');
    expect(emailInput.value).toBe('test@email.com');
  });

  it('updates password and repeat password fields', () => {
    render(
      <SignUpForm closeForm={mockCloseForm} toggleForm={mockToggleForm} />
    );

    const passwordInput = screen.getByPlaceholderText(
      'Password'
    ) as HTMLInputElement;
    const repeatPasswordInput = screen.getByPlaceholderText(
      'Repeat password'
    ) as HTMLInputElement;

    fireEvent.change(passwordInput, { target: { value: 'test_password' } });
    fireEvent.change(repeatPasswordInput, {
      target: { value: 'test_repeat_password' },
    });

    expect(passwordInput.value).toBe('test_password');
    expect(repeatPasswordInput.value).toBe('test_repeat_password');
  });

  it('toggles password visibility', () => {
    render(
      <SignUpForm closeForm={mockCloseForm} toggleForm={mockToggleForm} />
    );

    const passwordInput = screen.getByPlaceholderText(
      'Password'
    ) as HTMLInputElement;
    const toggleButton = screen.getByRole('button', { name: /show password/i });

    expect(passwordInput.type).toBe('password');

    fireEvent.click(toggleButton);

    expect(passwordInput.type).toBe('text');
  });

  it('toggles repeat password visibility', () => {
    render(
      <SignUpForm closeForm={mockCloseForm} toggleForm={mockToggleForm} />
    );

    const repeatPasswordInput = screen.getByPlaceholderText(
      'Repeat password'
    ) as HTMLInputElement;
    const toggleButton = screen.getByRole('button', {
      name: /show repeat password/i,
    });

    expect(repeatPasswordInput.type).toBe('password');

    fireEvent.click(toggleButton);

    expect(repeatPasswordInput.type).toBe('text');
  });

  it('calls closeForm when submitting', () => {
    render(
      <SignUpForm closeForm={mockCloseForm} toggleForm={mockToggleForm} />
    );

    const submitButton = screen.getByRole('button', { name: /Sign up/i });
    fireEvent.click(submitButton);
  });
});
