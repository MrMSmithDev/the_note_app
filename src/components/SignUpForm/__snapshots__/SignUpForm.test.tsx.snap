// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm component renders to match snapshot 1`] = `
<DocumentFragment>
  <form
    class="flex flex-col gap-5 bg-background dark:bg-background-dark rounded-md p-5"
  >
    <h3
      class="font-kanit font-bold text-2xl text-transparent bg-clip-text mr-auto bg-linear-to-tr from-blue-400 to-blue-600 py-2 select-none w-full text-center"
    >
      Sign up
    </h3>
    <div
      class="relative"
    >
      <svg
        class="absolute h-5 w-5 top-2 left-2"
        data-testid="user-icon"
        fill="none"
        viewBox="0 0 16 16"
      >
        <g
          stroke-width="0"
        />
        <g
          stroke-linecap="round"
          stroke-linejoin="round"
        />
        <g>
          <path
            d="M8 7C9.65685 7 11 5.65685 11 4C11 2.34315 9.65685 1 8 1C6.34315 1 5 2.34315 5 4C5 5.65685 6.34315 7 8 7Z"
            fill="#C8C8C9"
          />
          <path
            d="M14 12C14 10.3431 12.6569 9 11 9H5C3.34315 9 2 10.3431 2 12V15H14V12Z"
            fill="#C8C8C9"
          />
        </g>
      </svg>
      <label
        class="sr-only"
        for="username"
      />
      <input
        class="w-full p-2 rounded border-gray-200 border-2 border-solid dark:border-gray-800 focus:outline-gray-300 text-gray-800 dark:text-gray-200 tracking-wide pl-8"
        id="username"
        name="username"
        placeholder="Username"
        type="text"
        value=""
      />
    </div>
    <div
      class="relative"
    >
      <svg
        class="absolute h-5 w-5 top-2 left-2"
        data-testid="email-icon"
        fill="#156CFF"
        version="1.1"
        viewBox="0 -2.5 20 20"
      >
        <g
          stroke-width="0"
        />
        <g
          stroke-linecap="round"
          stroke-linejoin="round"
        />
        <g>
          <g
            fill="none"
            fill-rule="evenodd"
            stroke="none"
            stroke-width="1"
          >
            <g
              fill="#C8C8C9"
              transform="translate(-340.000000, -922.000000)"
            >
              <g
                transform="translate(56.000000, 160.000000)"
              >
                <path
                  d="M294,774.474 L284,765.649 L284,777 L304,777 L304,765.649 L294,774.474 Z M294.001,771.812 L284,762.981 L284,762 L304,762 L304,762.981 L294.001,771.812 Z"
                />
              </g>
            </g>
          </g>
        </g>
      </svg>
      <label
        class="sr-only"
        for="email"
      />
      <input
        class="w-full p-2 rounded border-gray-200 border-2 border-solid dark:border-gray-800 focus:outline-gray-300 text-gray-800 dark:text-gray-200 tracking-wide pl-8"
        id="email"
        name="email"
        placeholder="Email@example.com"
        type="email"
        value=""
      />
    </div>
    <div
      class="relative"
    >
      <button
        class="absolute h-5 w-5 top-2 left-2 cursor-pointer"
      >
        <span
          class="sr-only"
        >
          show password
        </span>
        <svg
          class=""
          data-testid="hide-password-icon"
          fill="none"
          viewBox="0 0 16 16"
        >
          <g
            stroke-width="0"
          />
          <g
            stroke-linecap="round"
            stroke-linejoin="round"
          />
          <g>
            <path
              clip-rule="evenodd"
              d="M16 16H13L10.8368 13.3376C9.96488 13.7682 8.99592 14 8 14C6.09909 14 4.29638 13.1557 3.07945 11.6953L0 8L3.07945 4.30466C3.14989 4.22013 3.22229 4.13767 3.29656 4.05731L0 0H3L16 16ZM5.35254 6.58774C5.12755 7.00862 5 7.48941 5 8C5 9.65685 6.34315 11 8 11C8.29178 11 8.57383 10.9583 8.84053 10.8807L5.35254 6.58774Z"
              fill="#C8C8C9"
              fill-rule="evenodd"
            />
            <path
              d="M16 8L14.2278 10.1266L7.63351 2.01048C7.75518 2.00351 7.87739 2 8 2C9.90091 2 11.7036 2.84434 12.9206 4.30466L16 8Z"
              fill="#C8C8C9"
            />
          </g>
        </svg>
      </button>
      <label
        class="sr-only"
        for="password"
      />
      <input
        class="w-full p-2 rounded border-gray-200 border-2 border-solid dark:border-gray-800 focus:outline-gray-300 text-gray-800 dark:text-gray-200 tracking-wide pl-8"
        id="password"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
    </div>
    <div
      class="relative"
    >
      <button
        class="absolute h-5 w-5 top-2 left-2 cursor-pointer"
      >
        <span
          class="sr-only"
        >
          show repeat password
        </span>
        <svg
          class=""
          data-testid="hide-password-icon"
          fill="none"
          viewBox="0 0 16 16"
        >
          <g
            stroke-width="0"
          />
          <g
            stroke-linecap="round"
            stroke-linejoin="round"
          />
          <g>
            <path
              clip-rule="evenodd"
              d="M16 16H13L10.8368 13.3376C9.96488 13.7682 8.99592 14 8 14C6.09909 14 4.29638 13.1557 3.07945 11.6953L0 8L3.07945 4.30466C3.14989 4.22013 3.22229 4.13767 3.29656 4.05731L0 0H3L16 16ZM5.35254 6.58774C5.12755 7.00862 5 7.48941 5 8C5 9.65685 6.34315 11 8 11C8.29178 11 8.57383 10.9583 8.84053 10.8807L5.35254 6.58774Z"
              fill="#C8C8C9"
              fill-rule="evenodd"
            />
            <path
              d="M16 8L14.2278 10.1266L7.63351 2.01048C7.75518 2.00351 7.87739 2 8 2C9.90091 2 11.7036 2.84434 12.9206 4.30466L16 8Z"
              fill="#C8C8C9"
            />
          </g>
        </svg>
      </button>
      <label
        class="sr-only"
        for="repeatPassword"
      />
      <input
        class="w-full p-2 rounded border-gray-200 border-2 border-solid dark:border-gray-800 focus:outline-gray-300 text-gray-800 dark:text-gray-200 tracking-wide pl-8"
        id="repeatPassword"
        name="repeatPassword"
        placeholder="Repeat password"
        type="password"
        value=""
      />
    </div>
    <button
      class="py-2 px-4 rounded-md font-quicksand font-bold text-gray-100 mr-auto bg-linear-to-tr from-blue-400 to-blue-600 w-full cursor-pointer"
      type="submit"
    >
      Sign up
    </button>
    <div
      class="text-center"
    >
      <p
        class="text-gray-800 dark:text-gray-200 text-xs"
      >
        Already have an account?
      </p>
      <button
        class="text-blue-400 text-sm font-bold cursor-pointer"
      >
        Sign in
      </button>
    </div>
  </form>
</DocumentFragment>
`;
